### DISCUSSION TITLE
Messages >0.5MB gets dropped occassionly when sent over SHM with QoS BEST_EFFORT

### DISCUSSION THREAD

**Discussion by dk-teknologisk-lag** [eProsima/Fast-DDS#4276] [Support]
Hey everyone

I'm currently working with interfacing various sensors to a mobile robot using ROS2, where we experience message drops. I have recently created a issue here: https://github.com/ros2/rmw_fastrtps/issues/739

But tried to go deeper and made small changes to the HelloWorldExample to use BEST_EFFORT and sending 10MB of messages.
It can be seen [here](https://github.com/eProsima/Fast-DDS/compare/master...dk-teknologisk-lag:Fast-DDS:bestefforthelloworld)

It seem to create SHM in /dev/shm, but only about 1MB. Also see the below screenshot for the samle lost callbacks, showing  messages to be lost.
![image](https://github.com/eProsima/Fast-DDS/assets/127200805/7ca852f4-8f6b-4c40-ac1f-55fdff6a6b5b)

I'm on a recent master and its build an run in a ubuntu 22.04 docker container.

Are there any thing I can tune when its over SHM?

I have searched for similar discussions, but they are often referring to communication across network.

Thanks, Lars

**Comment by Barry-Xu-2018** [comment]
Maybe you should modify example [HelloWorldExampleSharedMem](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/dds/HelloWorldExampleSharedMem) which is used for shared memory.

**Comment by dk-teknologisk-lag** [comment]
> Maybe you should modify example [HelloWorldExampleSharedMem](https://github.com/eProsima/Fast-DDS/tree/master/examples/cpp/dds/HelloWorldExampleSharedMem) which is used for shared memory.

Ahh, didn't see that. I have done so.

**Comment by fujitatomoya** [comment]
@dk-teknologisk-lag FYI, there are a couple of things i want to point out.
CC: @MiguelCompany @Barry-Xu-2018 @EduPonz (correct me if i got anything wrong)

- there are 2 features which use shared memory. (both creates the files under `/dev/shm` file system.)

  1. shared memory transport, https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/shared_memory/shared_memory.html (i think you are dealing with this one only.)
  2. data sharing delivery, https://fast-dds.docs.eprosima.com/en/latest/fastdds/transport/datasharing.html

> https://github.com/eProsima/Fast-DDS/compare/master...dk-teknologisk-lag:Fast-DDS:bestefforthelloworld

this uses `std::string` message type which is NOT bounded data type (Not PoD), that means this cannot use `data sharing delivery`. i think this is why @Barry-Xu-2018 introduced https://github.com/eProsima/Fast-DDS/discussions/4276#discussioncomment-8218106

> Messages >0.5MB gets dropped occassionly when sent over SHM with QoS BEST_EFFORT

and comes back to the main title, this sounds possible.
since using `Best Effort` QoS setting, there is always the possibility to drop the message.

**Comment by fujitatomoya** [comment]
i think we can iterate on https://github.com/ros2/rmw_fastrtps/issues/739 instead, so i will go ahead to close this discussion.
